src/main/java/com/eternalcode/discordapp/experience/listener/ExperienceVoiceListener.java

```diff
@@ -0,0 +1,75 @@
+package com.eternalcode.discordapp.experience.listener;
+
+import com.eternalcode.discordapp.config.ConfigManager;
+import com.eternalcode.discordapp.data.DataManager;
+import com.eternalcode.discordapp.experience.ExperienceConfig;
+import com.eternalcode.discordapp.experience.ExperienceService;
+import com.eternalcode.discordapp.experience.data.UsersVoiceActivityData;
+import net.dv8tion.jda.api.events.guild.voice.GuildVoiceUpdateEvent;
+import net.dv8tion.jda.api.hooks.ListenerAdapter;
+
+import java.time.Instant;
+import java.time.LocalTime;
+import java.time.ZoneId;
+
+public class ExperienceVoiceListener extends ListenerAdapter {
+
+    private final ExperienceConfig experienceConfig;
+
+    private final UsersVoiceActivityData usersVoiceActivityData;
+    private final DataManager dataManager;
+    private final ExperienceService experienceService;
+
+    public ExperienceVoiceListener(ExperienceConfig experienceConfig, UsersVoiceActivityData usersVoiceActivityData, DataManager dataManager, ExperienceService experienceService) {
+        this.experienceConfig = experienceConfig;
+        this.usersVoiceActivityData = usersVoiceActivityData;
+        this.dataManager = dataManager;
+        this.experienceService = experienceService;
+    }
+
+    @Override
+    public void onGuildVoiceUpdate(GuildVoiceUpdateEvent event) {
+        try {
+            if (event.getMember().getUser().isBot()) {
+                return;
+            }
+
+            this.leaveVoiceChannel(event);
+            this.joinVoiceChannel(event);
+
+            this.dataManager.save(this.usersVoiceActivityData);
+        }
+        catch (Exception exception) {
+            throw new RuntimeException(exception);
+        }
+    }
+
+    private void leaveVoiceChannel(GuildVoiceUpdateEvent event) {
+        if (event.getChannelLeft() != null) {
+            long userId = event.getMember().getIdLong();
+            this.usersVoiceActivityData.usersOnVoiceChannel.remove(event.getMember().getIdLong());
+            this.experienceService.addPoints(userId, this.calculatePoints(event));
+        }
```

Comment:
Nie ma tutaj potrzeby zamykania tej logiki w ife, dużo lepiej wydzielć to do dalejsze cześci metody, będzie to wtedy dużo bardziej czytelniejsze

```suggestion
        if (event.getChannelLeft() == null) {
        return
        }


                        long userId = event.getMember().getIdLong();
            this.usersVoiceActivityData.usersOnVoiceChannel.remove(event.getMember().getIdLong());
            this.experienceService.addPoints(userId, this.calculatePoints(event));
        
```

Reply:
Nie wiem czy czasem przy wywaleniu tego z if'a nie będzie ignorować joinChannel wtedy bo da return - według mnie lepiej by akurat to zostało w obecnej formie

Reply:
Dlaczego miałoby ignorować joinChannel? zrobiłem tylko invert ifa?

Reply:
No ale przecież ten return jest w warunku, to dlaczego miałbybyć zignorowany?